File: Backend/src/main/java/com/example/backend/service/ReviewService.java
package com.example.backend.service;

import com.example.backend.dto.request.CreateReviewRequest;
import com.example.backend.dto.response.ReviewResponse;
import com.example.backend.entity.Review;
import com.example.backend.entity.User;
import com.example.backend.exception.AppException;
import com.example.backend.exception.ErrorCode;
import com.example.backend.mapper.ReviewMapper;
import com.example.backend.repository.ReviewRepository;
import com.example.backend.repository.UserRepository;
import com.example.backend.repository.OrderRespository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class ReviewService {
    
    private final ReviewRepository reviewRepository;
    private final UserRepository userRepository;
    private final OrderRespository orderRepository;
    private final CloudinaryService cloudinaryService;
    private final ReviewMapper reviewMapper;
    
    /**
     * Tạo review mới cho seller
     */
    public ReviewResponse createReview(CreateReviewRequest request, String reviewerUsername) {
        // Tìm reviewer
        User reviewer = userRepository.findByUsername(reviewerUsername)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        
        // Tìm reviewee (seller)
        User reviewee = userRepository.findById(request.getRevieweeId())
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        
        // Kiểm tra xem reviewer đã review seller này chưa
        Review existingReview = reviewRepository.findByReviewerAndReviewee(reviewer, reviewee);
        if (existingReview != null) {
            throw new AppException(ErrorCode.REVIEW_ALREADY_EXISTS);
        }
        
        // Kiểm tra xem reviewer đã mua hàng thành công từ seller này chưa
        boolean hasCompletedPurchase = orderRepository.hasCompletedPurchase(reviewer.getId(), reviewee.getId());
        if (!hasCompletedPurchase) {
            throw new AppException(ErrorCode.PURCHASE_NOT_COMPLETED);
        }
        
        // Upload images nếu có
        List<String> imageUrls = new ArrayList<>();
        if (request.getImages() != null && !request.getImages().isEmpty()) {
            for (MultipartFile image : request.getImages()) {
                try {
                    String imageUrl = cloudinaryService.upload(image);
                    imageUrls.add(imageUrl);
                } catch (Exception e) {
                    log.error("Error uploading image: {}", e.getMessage());
                    throw new AppException(ErrorCode.UPLOAD_FAILED);
                }
            }
        }
        
        // Tạo review
        Review review = Review.builder()
                .reviewer(reviewer)
                .reviewee(reviewee)
                .rating(request.getRating())
                .comment(request.getComment())
                .imageUrls(imageUrls)
                .build();
        
        Review savedReview = reviewRepository.save(review);
        
        return reviewMapper.toReviewResponse(savedReview);
    }
    
    /**
     * Lấy tất cả review của một seller (cho seller đọc)
     */
    public Page<ReviewResponse> getReviewsForSeller(Long sellerId, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        User seller = userRepository.findById(sellerId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        Page<Review> reviews = reviewRepository.findByRevieweeOrderByCreatedAtDesc(seller, pageable);
        return reviews.map(reviewMapper::toReviewResponse);
    }
    
    /**
     * Lấy review theo rating cụ thể của một seller
     */
    public Page<ReviewResponse> getReviewsForSellerByRating(Long sellerId, Integer rating, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        Page<Review> reviews = reviewRepository.findByRevieweeIdAndRatingOrderByCreatedAtDesc(sellerId, rating, pageable);
        return reviews.map(reviewMapper::toReviewResponse);
    }
    
    /**
     * Lấy thống kê review của seller
     */
    public ReviewStatsResponse getReviewStats(Long sellerId) {
        Double averageRating = reviewRepository.getAverageRatingByRevieweeId(sellerId);
        Long totalReviews = reviewRepository.countByRevieweeId(sellerId);
        
        return ReviewStatsResponse.builder()
                .averageRating(averageRating != null ? averageRating : 0.0)
                .totalReviews(totalReviews)
                .build();
    }
    
    /**
     * Lấy review mà user đã viết
     */
    public Page<ReviewResponse> getReviewsByUser(Long userId, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        Page<Review> reviews = reviewRepository.findByReviewerOrderByCreatedAtDesc(user, pageable);
        return reviews.map(reviewMapper::toReviewResponse);
    }
    
    /**
     * Xóa review (chỉ reviewer mới có thể xóa)
     */
    public void deleteReview(Long reviewId, String username) {
        Review review = reviewRepository.findById(reviewId)
                .orElseThrow(() -> new AppException(ErrorCode.REVIEW_NOT_FOUND));
        
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
        
        // Chỉ reviewer mới có thể xóa review của mình
        if (review.getReviewer().getId() != user.getId()) {
            throw new AppException(ErrorCode.ACCESS_DENIED);
        }
        
        reviewRepository.delete(review);
    }
    
    // Inner class cho thống kê review
    @lombok.Data
    @lombok.Builder
    @lombok.NoArgsConstructor
    @lombok.AllArgsConstructor
    public static class ReviewStatsResponse {
        private Double averageRating;
        private Long totalReviews;
    }
}

File: Frontend/src/services/reviewService.js
import apiClient from "./apiClient";

// ==================== REVIEW SERVICE ====================
// Dịch vụ đánh giá: quản lý review của user về seller

const reviewService = {
  /**
   * Tạo review mới cho seller
   * @param {Object} reviewData - { revieweeId, rating, comment, images }
   * @param {string} username - Username của người tạo review
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async createReview(reviewData, username = null) {
    try {
      // Lấy username từ localStorage nếu không truyền vào
      if (!username) {
        const raw = localStorage.getItem("userData");
        if (raw) {
          try {
            const parsed = JSON.parse(raw);
            username =
              parsed?.username ||
              parsed?.user?.username ||
              parsed?.email ||
              parsed?.user?.email ||
              null;
          } catch (e) {
            return {
              success: false,
              message: "Không thể lấy thông tin người dùng từ localStorage",
            };
          }
        }
      }

      if (!username) {
        return {
          success: false,
          message: "Vui lòng đăng nhập trước khi đánh giá",
        };
      }

      // Tạo FormData cho multipart/form-data
      const formData = new FormData();
      formData.append("revieweeId", reviewData.revieweeId);
      formData.append("rating", reviewData.rating);
      if (reviewData.comment) {
        formData.append("comment", reviewData.comment);
      }
      
      // Thêm images nếu có
      if (reviewData.images && reviewData.images.length > 0) {
        reviewData.images.forEach((image) => {
          formData.append("images", image);
        });
      }

      const response = await apiClient.post(
        `/reviews/create?username=${encodeURIComponent(username)}`,
        formData,
        {
          headers: {
            "Content-Type": "multipart/form-data",
          },
        }
      );

      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Tạo review thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      // Thông báo lỗi thân thiện không có mã lỗi
      let userFriendlyMessage = 'Không thể gửi đánh giá';
      if (status === 401) {
        userFriendlyMessage = 'Phiên đăng nhập đã hết hạn. Vui lòng đăng nhập lại';
      } else if (status === 403) {
        userFriendlyMessage = 'Bạn không có quyền tạo đánh giá';
      } else if (status === 400) {
        userFriendlyMessage = backendMessage || 'Dữ liệu không hợp lệ. Vui lòng kiểm tra lại';
      } else if (status >= 500) {
        userFriendlyMessage = 'Hệ thống đang gặp sự cố. Vui lòng thử lại sau';
      } else if (!status) {
        userFriendlyMessage = 'Không thể kết nối đến máy chủ. Vui lòng kiểm tra kết nối internet';
      }
      
      return {
        success: false,
        message: userFriendlyMessage
      };
    }
  },

  /**
   * Lấy danh sách review của một seller
   * @param {number} sellerId - ID của seller
   * @param {number} page - Số trang (default: 0)
   * @param {number} size - Số item mỗi trang (default: 10)
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async getReviewsForSeller(sellerId, page = 0, size = 10) {
    try {
      const response = await apiClient.get(
        `/reviews/seller/${sellerId}?page=${page}&size=${size}`
      );
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Lấy danh sách review thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      // Thông báo lỗi thân thiện không có mã lỗi
      let userFriendlyMessage = 'Không thể tải đánh giá';
      if (status === 401) {
        userFriendlyMessage = 'Phiên đăng nhập đã hết hạn. Vui lòng đăng nhập lại';
      } else if (status === 403) {
        userFriendlyMessage = 'Bạn không có quyền xem đánh giá này';
      } else if (status === 404) {
        userFriendlyMessage = 'Không tìm thấy đánh giá';
      } else if (status >= 500) {
        userFriendlyMessage = 'Hệ thống đang gặp sự cố. Vui lòng thử lại sau';
      } else if (!status) {
        userFriendlyMessage = 'Không thể kết nối đến máy chủ. Vui lòng kiểm tra kết nối internet';
      }
      
      return {
        success: false,
        message: userFriendlyMessage
      };
    }
  },

  /**
   * Lấy danh sách review mà user đã viết
   * @param {number} userId - ID của user
   * @param {number} page - Số trang (default: 0)
   * @param {number} size - Số item mỗi trang (default: 10)
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async getReviewsByUser(userId, page = 0, size = 10) {
    try {
      const response = await apiClient.get(
        `/reviews/user/${userId}?page=${page}&size=${size}`
      );
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Lấy danh sách review thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      // Thông báo lỗi thân thiện không có mã lỗi
      let userFriendlyMessage = 'Không thể tải đánh giá';
      if (status === 401) {
        userFriendlyMessage = 'Phiên đăng nhập đã hết hạn. Vui lòng đăng nhập lại';
      } else if (status === 403) {
        userFriendlyMessage = 'Bạn không có quyền xem đánh giá này';
      } else if (status === 404) {
        userFriendlyMessage = 'Không tìm thấy đánh giá';
      } else if (status >= 500) {
        userFriendlyMessage = 'Hệ thống đang gặp sự cố. Vui lòng thử lại sau';
      } else if (!status) {
        userFriendlyMessage = 'Không thể kết nối đến máy chủ. Vui lòng kiểm tra kết nối internet';
      }
      
      return {
        success: false,
        message: userFriendlyMessage
      };
    }
  },

  /**
   * Xóa review
   * @param {number} reviewId - ID của review cần xóa
   * @param {string} username - Username của người xóa (phải là reviewer)
   * @returns {Promise<{success: boolean, message?: string}>}
   */
  async deleteReview(reviewId, username = null) {
    try {
      // Lấy username từ localStorage nếu không truyền vào
      if (!username) {
        const raw = localStorage.getItem("userData");
        if (raw) {
          try {
            const parsed = JSON.parse(raw);
            username =
              parsed?.username ||
              parsed?.user?.username ||
              parsed?.email ||
              parsed?.user?.email ||
              null;
          } catch (e) {
            return {
              success: false,
              message: "Không thể lấy thông tin người dùng từ localStorage",
            };
          }
        }
      }

      if (!username) {
        return {
          success: false,
          message: "Vui lòng đăng nhập trước khi xóa review",
        };
      }

      await apiClient.delete(
        `/reviews/${reviewId}?username=${encodeURIComponent(username)}`
      );

      return {
        success: true,
        message: "Xóa review thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      // Thông báo lỗi thân thiện không có mã lỗi
      let userFriendlyMessage = 'Không thể xóa đánh giá';
      if (status === 401) {
        userFriendlyMessage = 'Phiên đăng nhập đã hết hạn. Vui lòng đăng nhập lại';
      } else if (status === 403) {
        userFriendlyMessage = 'Bạn không có quyền xóa đánh giá này';
      } else if (status === 404) {
        userFriendlyMessage = 'Không tìm thấy đánh giá cần xóa';
      } else if (status >= 500) {
        userFriendlyMessage = 'Hệ thống đang gặp sự cố. Vui lòng thử lại sau';
      } else if (!status) {
        userFriendlyMessage = 'Không thể kết nối đến máy chủ. Vui lòng kiểm tra kết nối internet';
      }
      
      return {
        success: false,
        message: userFriendlyMessage
      };
    }
  },

  /**
   * Lấy thống kê review của seller
   * @param {number} sellerId - ID của seller
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async getReviewStats(sellerId) {
    try {
      const response = await apiClient.get(`/reviews/seller/${sellerId}/stats`);
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Lấy thống kê review thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      // Thông báo lỗi thân thiện không có mã lỗi
      let userFriendlyMessage = 'Không thể tải thống kê đánh giá';
      if (status === 401) {
        userFriendlyMessage = 'Phiên đăng nhập đã hết hạn. Vui lòng đăng nhập lại';
      } else if (status === 403) {
        userFriendlyMessage = 'Bạn không có quyền xem thống kê này';
      } else if (status === 404) {
        userFriendlyMessage = 'Không tìm thấy thống kê';
      } else if (status >= 500) {
        userFriendlyMessage = 'Hệ thống đang gặp sự cố. Vui lòng thử lại sau';
      } else if (!status) {
        userFriendlyMessage = 'Không thể kết nối đến máy chủ. Vui lòng kiểm tra kết nối internet';
      }
      
      return {
        success: false,
        message: userFriendlyMessage
      };
    }
  },

  /**
   * Lấy reviews của seller theo rating cụ thể
   * @param {number} sellerId - ID của seller
   * @param {number} rating - Rating (1-5)
   * @param {number} page - Số trang (default: 0)
   * @param {number} size - Số item mỗi trang (default: 10)
   * @returns {Promise<{success: boolean, data?: any, message?: string}>}
   */
  async getReviewsForSellerByRating(sellerId, rating, page = 0, size = 10) {
    try {
      const response = await apiClient.get(
        `/reviews/seller/${sellerId}/rating/${rating}?page=${page}&size=${size}`
      );
      const data = response?.data?.data ?? response?.data;
      return {
        success: true,
        data: data,
        message: response?.data?.message || "Lấy danh sách review theo rating thành công",
      };
    } catch (error) {
      const status = error?.response?.status;
      const backendMessage =
        error?.response?.data?.message || error?.message;
      // Thông báo lỗi thân thiện không có mã lỗi
      let userFriendlyMessage = 'Không thể tải đánh giá';
      if (status === 401) {
        userFriendlyMessage = 'Phiên đăng nhập đã hết hạn. Vui lòng đăng nhập lại';
      } else if (status === 403) {
        userFriendlyMessage = 'Bạn không có quyền xem đánh giá này';
      } else if (status === 404) {
        userFriendlyMessage = 'Không tìm thấy đánh giá';
      } else if (status >= 500) {
        userFriendlyMessage = 'Hệ thống đang gặp sự cố. Vui lòng thử lại sau';
      } else if (!status) {
        userFriendlyMessage = 'Không thể kết nối đến máy chủ. Vui lòng kiểm tra kết nối internet';
      }
      
      return {
        success: false,
        message: userFriendlyMessage
      };
    }
  },
};

export default reviewService;

